### PR TITLE
Update config module to use a class

### LINKED ISSUE #207: Retain more configuration process information in class
**Is your feature request related to a problem? Please describe.**
The `config` module currently defines a set of functions to work through the steps of reading and generating the configuration along with a final wrapper function `validate_config` that runs those functions in the correct sequence. This:
* only returns the config dictionary - all of the intermediate steps, such as the merged schema and the like are temporary internal variables
* conflates several actions, such as the output of combined configs.

**Describe the solution you'd like**
Restructure the `config` module to use a `Config` class, which retains information on:

* the input file paths
* the merged config
* the merged schema
* failed inputs, merge conflicts and config errors.

The new class would allow `Config.validate_config()` or something like it that includes more of the state of the configuration process.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR is an update of the `config` module. The previous code defined a suite of functions that were used as a chain to:

* resolve inputs file paths to individual files
* load TOML content from individual files
* combine content into a single config
* validate the config (and write a validated config file).

The chain of functions means that the config process is stateless - you only get the validated dictionary at the end. This PR introduces the `Config` class, which encapsulates the same steps into a class instance. This is a `dict` subclass, so `Config` instances behave in exactly the same way as the previous dictionary outputs from `validate_config`, but the instance also retains information about config sources and validation. It adds an `export_config` method to separate that functionality from validation.

Mostly this is just rearranging the existing function code into methods, but there are two things that need a closer look:

1. There is a bug fix 🐛 . The previous version used `dict(ChainMap(...)` to merge config contents across files. This works if the inputs all have different first level keys but does not recursively resolve contents:

    ```python
    In [27]: from collections import ChainMap
    
    In [28]: dict(ChainMap({'core': 'stuff'}, {'plant': 'stuff'}))
    Out[28]: {'plant': 'stuff', 'core': 'stuff'}
    
    In [29]: dict(ChainMap({'core': {'grid': 'stuff'}}, {'core': {'data': 'stuff'}}, {'plant': 'stuff'}))
    Out[29]: {'plant': 'stuff', 'core': {'grid': 'stuff'}}
    ```
    The existing implementation already included the `check_dict_leaves` function, which was a recursive search for duplicated dictionary paths across those inputs. This PR extends that as the `config_merge` function which piggybacks exactly the same recursion to resolve the recursive dictionary content at the same time as detection duplicates.

2. ~~The previous implementation explicitly built a merged schema for all of the models in a configuration that was then applied to the entire merged config. This doesn't seem necessary - the appropriate individual schemas can be applied to the merged config dictionary to validate each section in turn. I'm not quite so sure about this one - I've deliberately some functions that may be un-needed in place for the moment to make it easier to compare what is being dropped.~~ 
    This has now been reverted to using the merged schema approach.

Fixes #207

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by jacobcook1995 — APPROVED** [review]
The changes look broadly sensible to me.

My biggest concern is the amount of information being logged, from the tests it seems like the amount of detail given as to what has caused files not to be readable is pretty minimal. I this will become pretty hard to debug without significantly more information being made available as `JSON Schema` are easy to screw up.

Otherwise I just had a few queries

File: virtual_rainforest/core/config.py [hunk 1]
```diff
@@ -199,373 +169,452 @@ def get_schema(module_name: str, schema_file_path: Path) -> dict:
...
 
 
-def collect_files(cfg_paths: list[str]) -> list[Path]:
-    """Collect all toml files from a user specified list of files and directories.
+def config_merge(
+    dest: dict, source: dict, conflicts: tuple = (), path: str = ""
+) -> tuple[dict, tuple]:
+    """Recursively merge two dictionaries detecting duplicated key definitions.
+
+    This function returns a copy of the dest dictionary that has been extended
```
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/core/config.py:217]: One of the arguments - should be:

> input ``dest`` dictionary

File: virtual_rainforest/core/config.py [hunk 2]
```diff
@@ -199,373 +169,452 @@ def get_schema(module_name: str, schema_file_path: Path) -> dict:
...
+        # also should ensure that the config["core"]["modules"] element is populated
+        if self.config_errors:
+            return
+
+        # Generate a dictionary of schemas for requested modules
+        all_schemas: dict[str, Any] = {"core": core_schema}
+        requested_modules = self["core"]["modules"]
+        for module in requested_modules:
+            # Trap unknown model schemas
+            if module not in SCHEMA_REGISTRY:
```
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/core/config.py:494]: That's a good point. I'll look again at the schema merging - we've got the code and I'm not sure we gain anything massive by using the simpler approach. It isn't like the code is used more than once in a run?

File: tests/core/test_config.py [hunk 3]
```diff
@@ -6,448 +6,510 @@
...
+                (ERROR, "Config file path with non-TOML suffix"),
+                (CRITICAL, "Config paths not all valid: check log."),
+            ),
+            id="not_toml",
+        ),
+        pytest.param(
+            [".", "all_config.toml"],
+            pytest.raises(ConfigurationError),
+            (
+                (ERROR, "Repeated files in config paths:"),
```
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/test_config.py:204]: It's truncated. The `log_check` helper explicitly only validates the start of the string to avoid problems with changing paths etc. In this case, the real output is:

```
tests/core/test_config.py::test_Config_resolve_config_paths[dupes] [ERROR] - config - resolve_config_paths(376) - Repeated files in config paths: /private/var/folders/62/cz0ctby96lq7ggnytfbtd__40000gn/T/pytest-of-dorme/pytest-170/test_Config_resolve_config_pat3/data/all_config.toml
[CRITICAL] - config - resolve_config_paths(381) - Config paths not all valid: check log.
```

File: tests/core/test_config.py [hunk 4]
```diff
@@ -6,448 +6,510 @@
...
+                (
+                    ERROR,
+                    "Configuration error in ['core', 'grid', 'ny']: "
+                    "-10 is less than or equal to the minimum of 0",
+                ),
+                (CRITICAL, "Configuration contains schema violations: check log"),
             ),
+            id="minimum_value_violation",
         ),
+        # pytest.param(
```
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/test_config.py:418]: Huh - my bad. I think it wasn't actually an unexpected property? Will re-enable and check it out.

File: tests/core/test_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobcook1995) [reply tests/core/test_config.py:204]: Ahh right that makes sense (+ explains something I'd wondered about `log_check`)

File: tests/core/test_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobcook1995) [reply tests/core/test_config.py:418]: `soil.no_layers` definitely should be an unexpected property unless this branched of develop earlier than February

File: tests/core/test_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/test_config.py:204]: I think `log_check`  _has_ to work that way or you'll get run/platform specific errors.

File: virtual_rainforest/core/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/core/config.py:494]: I think I was trying to be clever - the single integrated validation seems cleaner and more proof against edge cases.

File: tests/core/test_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/test_config.py:418]: It branched off on Monday. I've killed a couple of old config revision branches that were hanging around.

File: tests/core/data/no_required_in_schema.json
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/data/no_required_in_schema.json:1]: Oh - OK. `no_properties` _is_ missing the model properties. `no_required` is just wrong. And the expected log messages were truncated, so didn't reveal the issue. Fixed now.

File: tests/core/test_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/core/test_config.py]: OK - that is now re-activated.

File: virtual_rainforest/core/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/core/config.py:501]: I've reverted to using the merged schema approach - I suspect it is more robust and it was very quick to reinstate that flow.

**Review by dalonsoa — COMMENTED** [review]
The changes seems sensible to me and, while not checked, I assume they work. But so did the previous version. Therefore, what I'm uncertain is the motivation: why `the instance also retains information about config sources and validation` is something needed? What's the use case of that information?

**Comment by davidorme** [comment]
@dalonsoa Yup - that isn't terribly clear.

* One existing example is separating the configuration writing out. The previous version only dumped the configuration file as part of validation, but the class now has a separate `export_config` method. You could of course just write the created dict to TOML, but I had in mind that the generated merged TOML file could include a commented header block giving the paths of the files it had been merged from. That extra information isn't implemented yet but in the future we are probably going to want to have a configuration tool that allows users to build and validate specific configurations from a library of files.

* Another is that in programmatic use, config errors and the like are stored within the class as objects that can be accessed, rather than that information just being logged.

**Review by dalonsoa — APPROVED** [review]
